### PR TITLE
Fix rgb_to_hls

### PR DESCRIPTION
**Pull request by Dipet**

### CODE REVIEW

File: kornia/color/hls.py [hunk 1]
```diff
@@ -148,4 +148,6 @@ def rgb_to_hls(image):
 
     h: torch.Tensor = (60 * hi) / 360  # hue
 
-    return torch.stack([h, l, s], dim=-3)
+    image = torch.stack([h, l, s], dim=-3)
+    image[torch.isnan(image)] = 0
```
Comment (edgarriba) [inline kornia/color/hls.py]: if is easier to add an epsilon value here:

```python
eps: float = 1e-8
denom: torch.Tensor = maxc + minc + eps
s: torch.Tensor = torch.where(l < 0.5, deltac / denom, deltac / (2 - denom))  # saturation
```

Comment (Dipet) [reply kornia/color/hls.py]: Nope. When I will add `epsilon = 1e-9` some pixels in return value will be different to OpenCV results.
In the code above with uint8 image difference in some pixel will be `255.`
```
def test(img):
    img = K.rgb_to_hls(img.float() / 255)
    img[0] *= 360

    img[0] /= 2
    img[1:] *= 255

    return img.round().clamp(0, 255).type(torch.uint8)

np.random.seed(0)

img = np.random.randint(0, 256, [1000, 1000, 3], np.uint8)

cv_img = cv.cvtColor(img, cv.COLOR_RGB2HLS)

k_img = torch.from_numpy(np.transpose(img, [2, 0, 1]))
k_img = test(k_img)
k_img = np.transpose(k_img.numpy(), [1, 2, 0])

print(np.abs(cv_img.astype(int) - k_img.astype(int)).max())
```

Comment (Dipet) [reply kornia/color/hls.py]: ok, I will do this soon

Comment (edgarriba) [reply kornia/color/hls.py]: @Dipet have you tried with images in floating point? opencv usually has slightly different implementation for different data types. Our scope is training meaning that we usually assume data coming in floating point precision

Comment (Dipet) [reply kornia/color/hls.py]: @edgarriba OpenCV uses epsilon for this conversion https://github.com/opencv/opencv/blob/a165f5557974c7ba32e5d6c966ec053d2bf932cc/modules/imgproc/src/color_hsv.simd.hpp#L688
So, I will implement it like in OpenCV and I will think how implement conversion in uint8.

File: kornia/color/hls.py [hunk 2]
```diff
@@ -105,7 +105,7 @@ def forward(self, image: torch.Tensor) -> torch.Tensor:  # type: ignore
         return rgb_to_hls(image)
 
 
-def rgb_to_hls(image):
+def rgb_to_hls(image, eps=1.19209e-07):
```
Comment (dvd42) [inline kornia/color/hls.py]: Typing

File: kornia/color/hls.py [hunk 3]
```diff
@@ -136,9 +136,10 @@ def rgb_to_hls(image):
 
     l: torch.Tensor = (maxc + minc) / 2  # luminance
 
-    deltac: torch.Tensor = maxc - minc
+    deltac: torch.Tensor = maxc - minc + eps
+    sumc = maxc + minc
```
Comment (dvd42) [inline kornia/color/hls.py]: Typing

File: kornia/color/hls.py [hunk 4]
```diff
@@ -148,4 +149,9 @@ def rgb_to_hls(image):
 
     h: torch.Tensor = (60 * hi) / 360  # hue
 
+    cond = deltac > eps
+    zeros = torch.zeros_like(h)
```
Comment (dvd42) [inline kornia/color/hls.py]: Typing

File: kornia/color/hls.py (hunk 2 above)
Comment (edgarriba) [reply kornia/color/hls.py]: Is this the only exact value that matches opencv ?

File: test/color/test_hls.py [hunk 5]
```diff
@@ -162,3 +166,17 @@ def op_script(data: torch.Tensor) -> torch.Tensor:
...
             assert_allclose(actual, expected)
+
+    def test_nan(self):
+        np.random.seed(0)
+
+        img = (np.random.randint(0, 256, [1000, 1000, 3], np.uint8) / 255.0).astype(np.float32)
+
+        cv_img = cv2.cvtColor(img, cv2.COLOR_RGB2HLS)
+
+        k_img = torch.from_numpy(np.transpose(img, [2, 0, 1]))
```
Comment (edgarriba) [inline test/color/test_hls.py]: FYI, we have `kornia.image_to_tensor` for this

File: test/color/test_hls.py [hunk 6]
```diff
@@ -162,3 +166,17 @@ def op_script(data: torch.Tensor) -> torch.Tensor:
...
+        k_img = kornia.rgb_to_hls(k_img.float())
+        k_img[0] *= 360
+        k_img = np.transpose(k_img.numpy(), [1, 2, 0])
```
Comment (edgarriba) [inline test/color/test_hls.py]: And `kornia.tensor_to_image`

**Comment by edgarriba** [comment]
@Dipet please check this PR, it turns out the implementation was somehow wrong https://github.com/kornia/kornia/pull/333

**Comment by Dipet** [comment]
Outdated

**Pull request closed without merging**